### PR TITLE
Replace 'with mesh' with 'with jax.set_mesh(mesh)'

### PR DESCRIPTION
**Pull request by SurbhiJainUSC**
# Description

This PR replaces the usage of the `with mesh` context manager with the newer `with jax.set_mesh(mesh)` API across the MaxText codebase. This change is being made to address JAX deprecation warnings.

Fixes: b/470443463

# Tests

SFT:
```
python3 -m maxtext.trainers.post_train.sft.train_sft \
model_name=llama3.1-8b-Instruct \
load_parameters_path=gs://maxtext-model-checkpoints/llama3.1_8b_instruct/2025-10-16/scanned/0/items \
run_name=$RUN_NAME \
base_output_directory=$BASE_OUTPUT_DIRECTORY \
steps=2
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Review by NuojCheng — APPROVED** [review]

**Review by khatwanimohit — APPROVED** [review]

**Review by bvandermoon — APPROVED** [review]

**Review by richjames0 — APPROVED** [review]
lgtm
